### PR TITLE
support flink sql.

### PR DESCRIPTION
**Pull request by gj-zhang**
1. flink sql supported.
2. table source support projection and filter push down.
3. table sink support upsert semantic. need define primary key.
4. refactor the connector config.

### CODE REVIEW

**Comment by link3280** [comment]
@gj-zhang Thanks a lot for your PR! It's a long-requested feature. I'll make time to review it in the coming week.

**Comment by gj-zhang** [comment]
ok. thank you.

**Comment by gj-zhang** [comment]
oh. sorry. this commit contain #8 pr.I don't know if I can submit this PR code together

**Comment by link3280** [comment]
@gj-zhang It would be great if you can split out the contents about #8. To accelerate the merge progress, I did a little adjustment to your previous PR, so it differs a little from your local branch. Maybe check out the master and cherry-pick the changes? Thanks!

**Comment by gj-zhang** [comment]
@link3280 i can't see any change at master branch.

**Comment by link3280** [comment]
@gj-zhang I mean if your commit is based on the same branch with #8 , then there might be some difference between the codes about #8 in master branch and your local branch. Anyway, we should exclude the changes about #8 to keep the PR concise.

**Comment by gj-zhang** [comment]
@link3280 ok. my brach is check out from master branch. and i merge #8 with copy code. i will remove it.

File: src/main/java/org/mongoflink/sink/MongoSink.java [hunk 1]
```diff
@@ -30,37 +28,38 @@
 
     private final MongoClientProvider clientProvider;
 
-    private DocumentSerializer<IN> serializer;
+    private final DocumentSerializer<IN> serializer;
 
-    private final SinkConfiguration configuration;
+    private final MongoConnectorOptions options;
 
-    public MongoSink(String connectionString,
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoSink.java:37]: As mentioned above, we'd better keep this constructor to be compatible.

File: src/main/java/org/mongoflink/sink/MongoCommitter.java [hunk 2]
```diff
@@ -24,11 +24,14 @@
  **/
 public class MongoCommitter implements Committer<DocumentBulk> {
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoCommitter.java:25]: These changes look like reverting #8. Can we exclude that?

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java [hunk 3]
```diff
@@ -33,7 +39,7 @@
 
     private transient MongoCollection<Document> collection;
 
-    private final ConcurrentLinkedQueue<Document> currentBulk = new ConcurrentLinkedQueue<>();
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoBulkWriter.java:36]: This should be from #8 as well.

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java [hunk 4]
```diff
@@ -49,30 +55,36 @@
 
     private final boolean flushOnCheckpoint;
 
+    private final boolean upsertEnable;
+    private final String[] upsertKeys;
+
     private final RetryPolicy retryPolicy = new RetryPolicy(3, 1000L);
 
-    private transient volatile boolean initialized = false;
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoBulkWriter.java:54]: We should keep `initialized ` to delay the schedule of flushes.

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java [hunk 5]
```diff
@@ -49,30 +55,36 @@
...
+        this.options = options;
+        if (!flushOnCheckpoint && this.options.getFlushInterval().getSeconds() > 0) {
             this.scheduler =
                     Executors.newScheduledThreadPool(
                             1, new ExecutorThreadFactory("mongodb-bulk-writer"));
             this.scheduledFuture =
                     scheduler.scheduleWithFixedDelay(
                             () -> {
                                 synchronized (MongoBulkWriter.this) {
-                                    if (initialized && !closed) {
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoBulkWriter.java:75]: ditto

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java [hunk 6]
```diff
@@ -204,18 +234,9 @@ private void rollBulkIfNeeded() {
     }
 
     private synchronized void rollBulkIfNeeded(boolean force) {
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoBulkWriter.java:236]: #8 code as well.

File: src/main/java/org/mongoflink/sink/MongoCommitter.java [hunk 7]
```diff
@@ -67,12 +77,6 @@ public List<DocumentBulk> commit(List<DocumentBulk> committables) throws IOExcep
 
     @Override
     public void close() throws Exception {
-        long deadline = System.currentTimeMillis() + TRANSACTION_TIMEOUT_MS;
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoCommitter.java:70]: #8 codes.

**Review by link3280 — CHANGES_REQUESTED** [review]
@gj-zhang Thanks a lot for the big PR! The codes look good in general, but with mainly two concerns:

1. Some lines in #8 are reverted accidentally as you mentioned. I guess you may override the codes with your local version when resolving conflicts. Please carefully exclude the unintentional changes.
2. The new options are great, but could we keep the codes about the old configuration at least for one or two versions (like we can add new methods rather than modify the existing ones)? Generally speaking, we should mark the old ones as deprecated first, and remove them after some versions.

Thanks again!

**Comment by gj-zhang** [comment]
@link3280 thanks for review.  I'm very busy these days, I will revise all of the above problems next week

**Comment by link3280** [comment]
@gj-zhang  No problem. Meanwhile, I'll fix the Github action, it seems to be broken.

File: src/main/java/org/mongoflink/config/MongoConnectorOptions.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
+package org.mongoflink.config;
+
+import java.io.Serializable;
+import java.time.Duration;
+
+/**
+ * mongo connector options
+ */
+public class MongoConnectorOptions implements Serializable {
```
[the remark this answers is not in the document]
Comment (gj-zhang) [reply src/main/java/org/mongoflink/config/MongoConnectorOptions.java:9]: done

File: src/main/java/org/mongoflink/serde/table/DocumentRowDataDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Override
+    public RowData deserialize(Document document) {
+        GenericRowData rowData = new GenericRowData(fieldNames.length);
+        for (int i = 0; i < fieldNames.length; i++) {
+            String fieldName = this.fieldNames[i];
+            Object o = document.get(fieldName);
+            if (o instanceof String) {
+                o = StringData.fromString((String) o);
```
[the remark this answers is not in the document]
Comment (gj-zhang) [reply src/main/java/org/mongoflink/serde/table/DocumentRowDataDeserializer.java]: please see org.apache.flink.table.data.RowData java doc.

File: src/main/java/org/mongoflink/serde/table/RowDataDocumentSerializer.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
+package org.mongoflink.serde.table;
+
+import org.apache.flink.formats.common.TimestampFormat;
+import org.apache.flink.formats.json.JsonFormatOptions;
+import org.apache.flink.formats.json.RowDataToJsonConverters;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonProcessingException;
```
[the remark this answers is not in the document]
Comment (gj-zhang) [reply src/main/java/org/mongoflink/serde/table/RowDataDocumentSerializer.java]: done.

File: src/main/java/org/mongoflink/sink/MongoSink.java (hunk 1 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoSink.java:37]: done

File: src/main/java/org/mongoflink/sink/MongoCommitter.java (hunk 2 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoCommitter.java]: done

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java (hunk 3 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoBulkWriter.java]: done

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java (hunk 4 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoBulkWriter.java:54]: done

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java (hunk 5 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoBulkWriter.java]: done

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java (hunk 6 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoBulkWriter.java]: done

File: src/main/java/org/mongoflink/sink/MongoCommitter.java (hunk 7 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoCommitter.java]: done

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java [hunk 11]
```diff
@@ -171,25 +182,68 @@ private synchronized void flush() {
...
+                            upserts.add(updateOneModel);
+                        }
+                        collection.bulkWrite(upserts, bulkWriteOptions);
+                    }
+
+                    iterator.remove();
+                    break;
+                } catch (MongoException e) {
+                    // maybe partial failure
+                    e.printStackTrace();
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoBulkWriter.java:240]: Please don't print to stderr

File: src/main/java/org/mongoflink/sink/MongoCommitter.java [hunk 12]
```diff
@@ -38,20 +38,30 @@ public class MongoCommitter implements Committer<DocumentBulk> {
             .writeConcern(WriteConcern.MAJORITY)
             .build();
 
+    private final boolean enableUpsert;
+    private final String[] upsertKeys;
+
     private final static long TRANSACTION_TIMEOUT_MS = 60_000L;
 
-    public MongoCommitter(MongoClientProvider clientProvider) {
+    public MongoCommitter(MongoClientProvider clientProvider, boolean enableUpsert, String[] upsertKeys) {
```
Comment (link3280) [inline src/main/java/org/mongoflink/sink/MongoCommitter.java:46]: Could we keep the old method and set the default value for the new parameters?

For example, 
```
public MongoCommitter(MongoClientProvider clientProvider) {
    this(clientProvider, false, new String[]{});
}
```

**Review by link3280 — COMMENTED** [review]
@gj-zhang Thank you very much for your excellent work! The codes look great! But there are two minor issues we need to fix before merging the PR, please take a look at my comments. Please forgive me for being picky.

File: src/main/java/org/mongoflink/sink/MongoBulkWriter.java (hunk 11 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoBulkWriter.java:240]: sorry for these code.

File: src/main/java/org/mongoflink/sink/MongoCommitter.java (hunk 12 above)
Comment (gj-zhang) [reply src/main/java/org/mongoflink/sink/MongoCommitter.java:46]: ok.

**Comment by gj-zhang** [comment]
@link3280 thank you very much for your review.

**Comment by link3280** [comment]
@gj-zhang Thanks a lot for your contributions! Would you like to join mongo-flink as a member? No additional responsibility is required, just as a recognition for your contributions to the project.

**Comment by gj-zhang** [comment]
> @gj-zhang Thanks a lot for your contributions! Would you like to join mongo-flink as a member? No additional responsibility is required, just as a recognition for your contributions to the project :)

ok. thank you.

**Comment by link3280** [comment]
@gj-zhang I've sent the invitation, please check your mailbox.

**Comment by madsjk816** [comment]
Could you provide this version on Maven as well? Thank you :)

**Pull request merged**